File: src/cleansia_android/core/src/main/java/cz/cleansia/core/auth/JwtDecoder.kt
package cz.cleansia.core.auth

import android.util.Base64
import org.json.JSONObject

/**
 * Lightweight JWT payload reader. We don't verify signatures client-side — the
 * server already did at issue time, and we trust our own token.
 */
object JwtDecoder {
    /** @return JWT `exp` claim as milliseconds-since-epoch, or null if unparseable. */
    fun extractExpiryMillis(jwt: String): Long? =
        payload(jwt)?.optLong("exp", -1L)?.takeIf { it > 0 }?.let { it * 1000L }

    /**
     * @return the user's email from the JWT. ASP.NET issues both the long
     * SOAP-style claim and a short "email" claim depending on config; check both.
     */
    fun extractEmail(jwt: String): String? {
        val json = payload(jwt) ?: return null
        return sequenceOf(
            "email",
            "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress",
        )
            .mapNotNull { key -> json.optString(key, "").takeIf { it.isNotBlank() } }
            .firstOrNull()
    }

    /**
     * @return the user's stable ID from the JWT (ASP.NET's `sub` or the SOAP nameidentifier
     * claim). Used for Sentry user context so crash reports can be correlated across
     * sessions without leaking PII.
     */
    fun extractUserId(jwt: String): String? {
        val json = payload(jwt) ?: return null
        return sequenceOf(
            "sub",
            "nameid",
            "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier",
        )
            .mapNotNull { key -> json.optString(key, "").takeIf { it.isNotBlank() } }
            .firstOrNull()
    }

    private fun payload(jwt: String): JSONObject? {
        val parts = jwt.split(".")
        if (parts.size != 3) return null
        return runCatching {
            val bytes = Base64.decode(parts[1], Base64.URL_SAFE or Base64.NO_PADDING)
            JSONObject(String(bytes))
        }.getOrNull()
    }
}

File: src/cleansia_ios/CleansiaCore/Sources/CleansiaCore/Auth/JwtDecoder.swift
import Foundation

public enum JwtDecoder {
    public static func expiry(of jwt: String) -> Date? {
        guard let exp = payload(of: jwt)?["exp"] else { return nil }
        let seconds: Double? = switch exp {
        case let value as Double: value
        case let value as Int: Double(value)
        case let value as NSNumber: value.doubleValue
        default: nil
        }
        guard let seconds, seconds > 0 else { return nil }
        return Date(timeIntervalSince1970: seconds)
    }

    public static func email(of jwt: String) -> String? {
        claim(
            of: jwt,
            keys: ["email", "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress"]
        )
    }

    public static func userId(of jwt: String) -> String? {
        claim(
            of: jwt,
            keys: [
                "sub",
                "nameid",
                "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier"
            ]
        )
    }

    private static func claim(of jwt: String, keys: [String]) -> String? {
        guard let payload = payload(of: jwt) else { return nil }
        for key in keys {
            if let value = payload[key] as? String, !value.isEmpty {
                return value
            }
        }
        return nil
    }

    private static func payload(of jwt: String) -> [String: Any]? {
        let parts = jwt.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 3, let data = base64URLDecode(String(parts[1])) else { return nil }
        return try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    private static func base64URLDecode(_ value: String) -> Data? {
        var base64 = value
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let remainder = base64.count % 4
        if remainder > 0 {
            base64 += String(repeating: "=", count: 4 - remainder)
        }
        return Data(base64Encoded: base64)
    }
}
